### PR TITLE
Update etcd client side to v3.4.1 in k8s 1.17

### LINKED ISSUE #82641: Update to etcd 3.4 in Kubernetes 1.17 release
follow the same pattern that we updated to 3.3.15:
- client side changes needs to be vendored in (similar to https://github.com/kubernetes/kubernetes/pull/81434)
- update default etcd server (similar to https://github.com/kubernetes/kubernetes/pull/82199)

### PR DESCRIPTION
**Pull request by wenjiaswe**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Update etcd version to v3.4 early in k8s 1.17

**Which issue(s) this PR fixes**:

Fixes #82641

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Update etcd client side to v3.4.1
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

/cc @dims @liggitt  @gyuho @jingyih @jpbetz 
/sig api-machinery

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -406,13 +412,16 @@ replace (
...
+	github.com/urfave/cli => github.com/urfave/cli v1.20.0
 	github.com/urfave/negroni => github.com/urfave/negroni v1.0.0
 	github.com/vishvananda/netlink => github.com/vishvananda/netlink v0.0.0-20171020171820-b2de5d10e38e
 	github.com/vishvananda/netns => github.com/vishvananda/netns v0.0.0-20171111001504-be1fbeda1936
 	github.com/vmware/govmomi => github.com/vmware/govmomi v0.20.1
 	github.com/xiang90/probing => github.com/xiang90/probing v0.0.0-20160813154853-07dd2e8dfe18
 	github.com/xlab/handysort => github.com/xlab/handysort v0.0.0-20150421192137-fb3537ed64a1
 	github.com/xordataexchange/crypt => github.com/xordataexchange/crypt v0.0.3-0.20170626215501-b2862e3d0a77
+	go.etcd.io/bbolt => go.etcd.io/bbolt v1.3.3
+	go.etcd.io/etcd => go.etcd.io/etcd v0.0.0-20190917205325-a14579fbfb1a
```
Comment (jpbetz) [inline go.mod]: Should be the `v3.4.1` version here?

File: staging/src/k8s.io/apiextensions-apiserver/go.mod [hunk 2]
```diff
@@ -35,13 +35,27 @@ require (
...
+	go.uber.org/zap => go.uber.org/zap v0.0.0-20180814183419-67bc79d13d15
 	golang.org/x/crypto => golang.org/x/crypto v0.0.0-20181025213731-e84da0312774
 	golang.org/x/lint => golang.org/x/lint v0.0.0-20181217174547-8f45f776aaf1
+	golang.org/x/net => golang.org/x/net v0.0.0-20190812203447-cdfb69ac37fc
 	golang.org/x/oauth2 => golang.org/x/oauth2 v0.0.0-20190402181905-9f3314589c9a
 	golang.org/x/sync => golang.org/x/sync v0.0.0-20181108010431-42b317875d0f
 	golang.org/x/sys => golang.org/x/sys v0.0.0-20190209173611-3b5209105503
 	golang.org/x/text => golang.org/x/text v0.3.1-0.20181227161524-e6919f6577db
 	golang.org/x/time => golang.org/x/time v0.0.0-20161028155119-f51c12702a4d
+	google.golang.org/grpc => google.golang.org/grpc v1.23.0
```
Comment (jpbetz) [inline staging/src/k8s.io/apiextensions-apiserver/go.mod]: `v1.23.1` ? I see some updates to `v1.23.1`.  There are a bunch of other `v1.23.0` usages below that all need to be updated too.

**Comment by jpbetz** [comment]
For the build error: 
> Deprecated pkg/transport.TLSInfo.CAFile field. Use pkg/transport.TLSInfo.TrustedCAFile instead (CAFile field has been marked deprecated since v2.1).

xref https://github.com/etcd-io/etcd/blob/master/CHANGELOG-3.4.md#breaking-changes

File: staging/src/k8s.io/apiextensions-apiserver/go.mod (hunk 2 above)
Comment (wenjiaswe) [reply staging/src/k8s.io/apiextensions-apiserver/go.mod]: Yes, I just updated the lint-dependencies, the first push was done before that. Now it's v1.23.1.

File: go.mod (hunk 1 above)
Comment (wenjiaswe) [reply go.mod]: @liggitt, based on your comments here, https://github.com/kubernetes/kubernetes/pull/81434#issuecomment-522394787, is this intended?

Comment (liggitt) [reply go.mod]: unfortunately, yes. etcd (like the k8s.io libraries) is using a semver tag >1.x with a go.mod file, and is not using semantic import versioning (e.g. naming their module `go.etcd.io/etcd/v3`)

go considers that invalid, and that means go will not honor/preserve the semver tag and will resolve the version to a pseudo-version that reads as `v0.0.0-<commit-date>-<sha>`

File: go.mod [hunk 3]
```diff
@@ -38,6 +38,7 @@ require (
 	github.com/containerd/typeurl v0.0.0-20190228175220-2a93cfde8c20 // indirect
 	github.com/containernetworking/cni v0.7.1
 	github.com/coredns/corefile-migration v1.0.2
+	github.com/coreos/bbolt v1.3.1-coreos.6 // indirect
 	github.com/coreos/etcd v3.3.15+incompatible
```
Comment (liggitt) [inline go.mod]: I expected to see us drop `github.com/coreos/etcd` as a dependency... is there a reason we did not?

Comment (wenjiaswe) [reply go.mod]: yes, I am going to update go.mod, haven't pushed it to this PR yet, as it's still WIP. I am still finding my way around updating vendor. Should I resolve building error with updated etcd 3.4.1 first, or should I remove the old github.com/coreos/etcd from go.mod first? After I remove  github.com/coreos/etcd from go.mod, do I need to update-vendor.sh again?

**Comment by wenjiaswe** [comment]
@liggitt would you please help to take a look at dependencies error? I keep removing github.com/coreos/etcd from go.mod, but then this test failed with following error and if I do update-vendor.sh, it will add back github.com/coreos/etcd. Is there anything that I missed?

Your go.mod file is different:
--- go.mod	2019-09-27 09:08:41.688631116 +0000
+++ /home/prow/go/src/k8s.io/kubernetes/_tmp/kube-vendor.BpEXHk/kubernetes/go.mod	2019-09-27 09:10:35.772244027 +0000
@@ -228,6 +228,7 @@ replace (
 	github.com/containerd/typeurl => github.com/containerd/typeurl v0.0.0-20190228175220-2a93cfde8c20
 	github.com/containernetworking/cni => github.com/containernetworking/cni v0.7.1
 	github.com/coredns/corefile-migration => github.com/coredns/corefile-migration v1.0.2
+	github.com/coreos/etcd => github.com/coreos/etcd v3.3.10+incompatible
 	github.com/coreos/go-etcd => github.com/coreos/go-etcd v2.0.0+incompatible
 	github.com/coreos/go-oidc => github.com/coreos/go-oidc v2.1.0+incompatible
 	github.com/coreos/go-semver => github.com/coreos/go-semver v0.3.0
Vendor Verify failed.
If you're seeing this locally, run the below command to fix your go.mod:
hack/update-vendor.sh

**Comment by dims** [comment]
@wenjiaswe looks like its coming from viper. It should be ok to add that line back in. we are ok as long as we don'g vendor in the files from github.com/coreos/etcd into k/k for sure.

```
vendor/github.com/spf13/viper/go.mod:	github.com/coreos/etcd v3.3.10+incompatible // indirect
vendor/github.com/spf13/viper/go.sum:github.com/coreos/etcd v3.3.10+incompatible/go.mod h1:uF7uidLiAD3TWHmW31ZFd/JWoc32PjwdhPthX9715RE=
vendor/github.com/spf13/cobra/go.sum:github.com/coreos/etcd v3.3.10+incompatible/go.mod h1:uF7uidLiAD3TWHmW31ZFd/JWoc32PjwdhPthX9715RE=
```

**Comment by wenjiaswe** [comment]
@wojtek-t All tests passed, the PR is ready for review. Would you please help to take a look? Thanks!

**Comment by wojtek-t** [comment]
This looks reasonable to me. My only comment would be:
@wenjiaswe - can you please squash the commits modifying vendor/modules into a single commit?

@liggitt - do you want to take a look into go modules stuff one more time?

File: Godeps/LICENSES [hunk 4]
```diff
@@ -6889,6 +6651,35 @@ Apache License
...
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+
+<http://www.opensource.org/licenses/mit-license.php>
```
Comment (liggitt) [inline Godeps/LICENSES]: cc @cblecker for ack of license

File: go.mod [hunk 5]
```diff
@@ -219,22 +220,23 @@ replace (
...
 	github.com/codegangsta/negroni => github.com/codegangsta/negroni v1.0.0
 	github.com/container-storage-interface/spec => github.com/container-storage-interface/spec v1.1.0
 	github.com/containerd/console => github.com/containerd/console v0.0.0-20170925154832-84eeaae905fa
 	github.com/containerd/containerd => github.com/containerd/containerd v1.0.2
 	github.com/containerd/typeurl => github.com/containerd/typeurl v0.0.0-20190228175220-2a93cfde8c20
 	github.com/containernetworking/cni => github.com/containernetworking/cni v0.7.1
 	github.com/coredns/corefile-migration => github.com/coredns/corefile-migration v1.0.2
-	github.com/coreos/bbolt => github.com/coreos/bbolt v1.3.1-coreos.6
-	github.com/coreos/etcd => github.com/coreos/etcd v3.3.15+incompatible
+	github.com/coreos/etcd => github.com/coreos/etcd v3.3.10+incompatible
```
Comment (liggitt) [inline go.mod]: if we're still going to list this, let's keep it at the same level we were at previously (v3.3.15)

Comment (wenjiaswe) [reply go.mod]: @liggitt, I tried to keep github.com/coreos/etcd at v3.3.15, but the verify-vendor keeps wanting me to pin it back to 3.3.10. 

```
github.com/coreos/etcd
    pinned:    v3.3.15+incompatible
    preferred: v3.3.10+incompatible
    hack/pin-dependency.sh github.com/coreos/etcd v3.3.10+incompatible
```

I wonder if this is related to what you mentioned in here: https://github.com/kubernetes/kubernetes/pull/81434#issuecomment-522673849

```
github.com/spf13/viper@v1.3.2 github.com/coreos/go-etcd@v2.0.0+incompatible
github.com/spf13/viper@v1.3.2 github.com/coreos/etcd@v3.3.10+incompatible
```
How or where could I see this and how to resolve this "transitive dependencies on any semantically-tagged etcd versions."?

**Comment by liggitt** [comment]
> I tried to keep github.com/coreos/etcd at v3.3.15, but the verify-vendor keeps wanting me to pin it back to 3.3.10.

ah, ok. we don't depend on it directly, so our expressed version gets dropped, then the effective version comes from our transitive dependency. it's fine to take the resolved version, then

File: go.mod [hunk 6]
```diff
@@ -301,9 +303,9 @@ replace (
 	github.com/gorilla/mux => github.com/gorilla/mux v1.7.0
 	github.com/gorilla/websocket => github.com/gorilla/websocket v1.4.0
 	github.com/gregjones/httpcache => github.com/gregjones/httpcache v0.0.0-20170728041850-787624de3eb7
-	github.com/grpc-ecosystem/go-grpc-middleware => github.com/grpc-ecosystem/go-grpc-middleware v0.0.0-20190222133341-cfaf5686ec79
+	github.com/grpc-ecosystem/go-grpc-middleware => github.com/grpc-ecosystem/go-grpc-middleware v1.0.1-0.20190118093823-f849b5445de4
```
[the remark this answers is not in the document]
Comment (wenjiaswe) [reply go.mod]: Yes, this is suggested by lint-dependency.sh. I don't know why though.

```
github.com/grpc-ecosystem/go-grpc-middleware
    pinned:    v0.0.0-20190222133341-cfaf5686ec79
    preferred: v1.0.1-0.20190118093823-f849b5445de4
    hack/pin-dependency.sh github.com/grpc-ecosystem/go-grpc-middleware v1.0.1-0.20190118093823-f849b5445de4

```

**Comment by wenjiaswe** [comment]
@wojtek-t @liggitt Comments addressed, would you please take another look? Thanks!

**Comment by wojtek-t** [comment]
LGTM for me, but I think Jordan should be the one approving this.

**Comment by wenjiaswe** [comment]
Thanks @wojtek-t , I will wait for Jordan.
/assign @liggitt

**Comment by liggitt** [comment]
status: still trying to work through the grpc dependency backleveling to understand if it is safe or if we need to find a way to stay at our current level.

There are also two in-flight issues related to etcd (https://github.com/kubernetes/kubernetes/issues/83028 and https://github.com/kubernetes/kubernetes/issues/83550) that we may want to resolve in the 3.3.x stream for picking to 1.16.x before we bump this.

**Comment by wenjiaswe** [comment]
Thanks @liggitt , I will keep an eye on those two etcd related issues and push to merge the fix if possible.

**Comment by wenjiaswe** [comment]
Now that etcd v3.3.16 and v3.4.2 is released. We will update etcd client and server from 3.3.15 to 3.3.16 for k8s 1.16, rebase, then update etcd to 3.4.2 for k8s 1.17. I will wait till that's done and resume the work here. 

/cc @jpbetz @liggitt @wojtek-t Please correct me if I am wrong.

**Comment by liggitt** [comment]
that sounds right

**Comment by wenjiaswe** [comment]
Wait for #83915 and #83804

**Comment by wenjiaswe** [comment]
Close this PR and update etcd to v3.4.2 in #83987

**Pull request closed without merging**